rust/rope/src/delta.rs

```diff
@@ -63,20 +63,55 @@ impl<N: NodeInfo> Delta<N> {
 
     /// If this delta represents a simple insertion, returns the inserted node.
     pub fn as_simple_insert(&self) -> Option<&Node<N>> {
-        if self.els.len() <= 3 && self.iter_deletions().count() == 0 {
-            for el in &self.els {
-                if let &DeltaElement::Insert(ref n) = el {
-                    return Some(n)
+        let mut iter = self.els.iter();
+        let mut el = iter.next();
+        let mut i = 0;
+        if let Some(&DeltaElement::Copy(beg, end)) = el {
+            if beg != 0 {
+                return None;
+            }
+            i = end;
+            el = iter.next();
+        }
+        if let Some(&DeltaElement::Insert(ref n)) = el {
+            el = iter.next();
+            if el.is_none() {
+                if i == self.base_len {
+                    return Some(n);
+                }
+            } else if let Some(&DeltaElement::Copy(beg, end)) = el {
+                if i == beg && end == self.base_len && iter.next().is_none() {
+                    return Some(n);
                 }
             }
         }
         None
     }
 
     /// Returns `true` if this delta represents a single deletion without
-    /// any insertions.
+    /// any insertions. Note that this is `false` for the trivial delta.
     pub fn is_simple_delete(&self) -> bool {
-        self.els.len() <= 2 && self.iter_inserts().count() == 0
+        if self.els.is_empty() && self.base_len > 0 {
+            return true;
+        }
+        if let DeltaElement::Copy(beg, end) = self.els[0] {
+            if beg == 0 {
+                if self.els.len() == 1 {
+                    // Deletion at end
+                    end < self.base_len
```

Comment:
isn't this an invariant?

Reply:
See above, not if the delta is trivial and base_len is nonzero.